File: ios/LocalStorageModule.swift
//
//  LocalStorageModule.swift
//  RNLocalStorage
//
//  Created by Rene Corrales on 11/3/23.
//

import React
import Foundation

struct Todo: Codable {
  let id: Int
  let title: String
  let userId: Int
  let completed: Bool
}

@objc(LocalStorageModule)
class LocalStorageModule: NSObject {
  
  /// Set vars as constants in react native modules
  @objc func constantsToExport() -> [AnyHashable: Any]! {

    return [
      "str": "Hello from modules",
      "num": 18,
      "bool": true,
      "null": NSNull(),
      "float": 14.4,
      "obj": ["a": "a", "b": "b"],
      "arr": ["item1", "item2"],
    ]
  }


  /// Necessary to instance module without constructor
  @objc static func requiresMainQueueSetup() -> Bool {
    return true
  }
  
  @objc func getTodos(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) -> Void {
    do {
      let todos = try self.fetchTodos()
      
      let todosMap = self.todosToMap(todos)
      
      resolve(todosMap)
    } catch {
      reject("Internal", "Error obtain", error)
    }
  }
  
  private func todosToMap(_ todos: [Todo]) -> [[String: Any]] {
    return todos.map { todo in
      return  [
          "id": todo.id,
          "title": todo.title,
          "userId": todo.userId,
          "completed": todo.completed
        ]
      
    }
  }
  
  private func fetchTodos() throws -> [Todo] {
    let url = URL(string: "https://jsonplaceholder.typicode.com/todos")
    let semaphore = DispatchSemaphore(value: 0)
    var result: Data = Data()
    
       
    let task = URLSession.shared.dataTask(with: url!) {(data, response, error) in
      result.append(data!)
      semaphore.signal()
    }
       
    task.resume()
    semaphore.wait()
    let jsonDecoder = JSONDecoder()
    return try jsonDecoder.decode([Todo].self, from: result)
      
  }
  
  @objc func getTodosCb(_ callback: RCTResponseSenderBlock) {
    do {
      let todos = try self.fetchTodos()
      
      let todosMap = self.todosToMap(todos)
      
      callback([NSNull(), todosMap])
    } catch {
      callback([error.localizedDescription, NSNull()])
    }
  }
  
  @objc func sayHello(_ name: String) -> String {
     return "hello \(name)"
  }
}

File: android/app/src/main/java/com/rnlocalstorage/NativeModules/LocalStorageModule.kt
package com.rnlocalstorage.NativeModules



import com.facebook.common.internal.Throwables
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Callback
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.WritableNativeArray

import kotlinx.serialization.*
import kotlinx.serialization.json.*
import java.net.URL


@Serializable
class Todo(val id: Int, val userId: Int, val title: String, val completed: Boolean)

class LocalStorageModule(reactContext: ReactApplicationContext): ReactContextBaseJavaModule(reactContext) {
    override fun getName(): String = "LocalStorageModule"

    override fun getConstants(): MutableMap<String, Any?> {
        val constants = mutableMapOf<String, Any?>()
        constants["str"] = "Hello from native modules"
        constants["num"] = 12
        constants["float"] = 12.5
        constants["bool"] = true
        constants["obj"] = mutableMapOf("a" to "a", "b" to "b")
        constants["arr"] = arrayOf("item1", "item2")
        constants["null"] = null

        return constants
    }

    @ReactMethod
    fun getTodos(promise: Promise) {
        try {
            val todos = this.fetchTodos()
            val todosWritableNativeArray = this.makeWritableNativeTodo(todos)

            promise.resolve(todosWritableNativeArray)
        } catch (e: Exception) {

            promise.reject(e)
        }

    }

    private fun fetchTodos(): Array<Todo> {
        val rawData = URL("https://jsonplaceholder.typicode.com/todos").openStream().bufferedReader().use { it.readText() }
        return Json.decodeFromString(rawData)

    }

    private fun todoToMap(todo: Todo): MutableMap<String, Any> {
        return mutableMapOf(
            "id" to todo.id,
            "title" to todo.title,
            "userId" to todo.userId,
            "completed" to todo.completed
        )
    }

    private fun makeWritableNativeTodo(todos: Array<Todo>): WritableNativeArray? {
        val todosListNativeMap = todos.map {
            Arguments.makeNativeMap(
                todoToMap(it)
            )
        }

        return Arguments.makeNativeArray(todosListNativeMap)
    }

    @ReactMethod
    fun getTodosCb(callback: Callback) {
        try {
            val todos = this.fetchTodos()
            val todosWritableNativeArray = this.makeWritableNativeTodo(todos)

            callback.invoke(null, todosWritableNativeArray)
        } catch (e: Exception) {
            callback.invoke(e.message, null)
        }
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    fun sayHello(name: String): String {
        return "hello $name"
    }

    override fun hasConstants(): Boolean {
        return true
    }

}
